### ISSUE TITLE
handle 'too-many-instance-attributes' differently for dataclass

### ISSUE THREAD

**Issue by vince6e74** [pylint-dev/pylint#9058]
### Current problem

Dataclasses sometimes deliberately have a lot of arguments, which is also the purpose of the classes. With normal classes, methods and functions it makes sense if the argument list is not too long. 

### Desired solution

Therefore a conscious distinction between dataclass and normal class would be meaningful with the argument number.

**Comment by nickdrozd** [comment]
@vince6e74 Could you give an example of this?

**Comment by anton-ubi** [comment]
I presume an example describing what the OP means could be:

```python
# pylint: disable=missing-class-docstring,missing-module-docstring

from dataclasses import dataclass


@dataclass
class Foo:
    one: int
    two: int
    three: int
    four: int
    five: int
    six: int
    seven: int
    eight: int

```

Output:

```
$ > pylint test.py           
************* Module test
test.py:7:0: R0902: Too many instance attributes (8/7) (too-many-instance-attributes)

------------------------------------------------------------------
Your code has been rated at 9.00/10 (previous run: 0.00/10, +9.00)
```

I've encountered the same situation myself. I would have expected that dataclasses are excluded from the R0902 check.

**Comment by clo-vis** [comment]
…while with
```python
class Foo(NamedTuple):
    one: int
    two: int
    three: int
    four: int
    five: int
    six: int
    seven: int
    eight: int
    nine: int
    ten: int
    eleven: int
    twelve: int
    thirteen: int
    fourteen: int
    fifteen: int
```
pylint is totally happy and does not complain at all.

**Comment by pushpendra-tripathi** [comment]
I'm finding this issue frustrating because data classes are intended to store data, and the default limit of 7 arguments seems too restrictive for this use case. It would be helpful to have an option to increase this limit specifically for data classes, while still maintaining the default limit of 7 arguments for regular classes.

**Comment by nickurak** [comment]
I started looking into this, and pretty quickly ran into a design question:

It makes sense to me to allow a dataclass to have 10+ instance attributes, in the same way that a NamedTuple could have a bunch (*). But what if you had a dataclass and added a handful of methods? It starts to look a lot more like a normal class in that case, with some convenience functionality around field management.


(*) I'm not entirely sure that pylint's ignoring of a large number NamedTuple "instance attributes" is actually intentional -- in checkers/design_analysis.py:visit_classdef, they're not explicitly skipped due to the NamedTuple-ness of the class, but either because they're not in the astroid node.instance_attrs or because it's filtered out in the `filtered_attrs` list comprehension.

**Comment by nickurak** [comment]
Yeah, astroid doesn't count a NamedTuple's instance attributes as attributes (ie, include them in node.instance_attrs). It would be helpful to have a statement from the maintainers on whether that's the desired behavior for NamedTuples, or just an accident, or a bug.

**Comment by Pierre-Sassoulas** [comment]
It's probably an accident that named tuple and dataclasses are treated differently. 

My gut reaction is to agree that this need to be consistent. Two possibilities feels consistent to me right now but I didn't think long about it: 
1) allow more than 7 instances attribute for NamedTuple or dataclasse without any methods (but still disallow when there are methods in the dataclass)
2) disallow more than 7 instances attributes for both NamedTuple and dataclasses (in the spirit of separating concepts even in data classes) 

This discussion will be opinionated and we'll probably need to add option to accommodate all opinions in the end but we have to define the default behavior in any case. I'm personally in favor of 1).

**Comment by janmondry** [comment]
It would be useful to be able to configure a separate `--max-attributes` limit for dataclasses, even when deciding on option 1) proposed above -- sometimes dataclasses need just one or two methods, but still have lots of arguments.
